### ISSUE TITLE
Can't use `AsyncLocalStorage.getItem` with `.concat`

### ISSUE THREAD

**Issue by supdegrave** [cyrilletuzi/angular-async-local-storage#25]
Hi, I'm trying to implement a system that manages an in-memory cache, backed by IndexedDB, populated by resources on a CDN. I want to return an observable that contains the data, as follows: 

* if the data is in the in-memory cache, return it in an Observable
* else if the data is in IndexedDb, return read Observable
* else return http.get Observable (and write to IDB/cache)

It seems I should be able to combine the IndexedDB read call and the http.get request via `concat`, and return the `first` non-null response, implemented roughly as follows (this is simplified, and subscribed elsewhere): 

```
get(path): Observable<any> {
    const data = this._cache[path];
    const uri = this.getCdnPath(path);

    // storageService.read wraps AsyncLocalStorage.getItem
    const readFromStorage$ = this.storageService.read(path);

    // Angular HttpClient
    const httpGet$ = this.http.get(uri); 

    return (data) 
        ? of(data) 
        : concat(readFromStorage$, httpGet$).pipe(first(result => result));
}
```

This doesn't proceed past the `storageService.read` call. I suspect that getFile isn't returning complete, so the http.get call is abandoned (from RxJs documentation for `concat`): 

> Note that if some input Observable never completes, concat will also never complete and Observables following the one that did not complete will never be subscribed. 

If I reverse the order inside `.concat`, both calls do proceed as expected. 

NOTE: I tried to set up a plunker for this but was getting import errors.

**Comment by supdegrave** [comment]
Looking at the `.getItem` source, I suspect that the problem is here: 

```
/* Merging success and errors events and autoclosing the observable */
return (observableRace(success, this.toErrorObservable(request, `getter`)) as Observable<T | null>)
    .pipe(first());
```

**Comment by cyrilletuzi** [comment]
The `.pipe(first())` is explicitly done for the observable to complete, so it's weird.

I'll look into this, but I'll need some time.

**Comment by cyrilletuzi** [comment]
Did you take note that : 

> Not finding an item is not an error, it succeeds but returns null.

? So if there is no data in local storage, and as you do `.pipe(first())` just after, it seems normal it does not get into the http call.

**Comment by cyrilletuzi** [comment]
Not sure the operators are the best ones here, but the logic of what you want to do may be more like :

```typescript
return of(data)
  .switchMap((value) => (value != undefined) ? of(value) : readFromStorage$)
  .switchMap((value) => (value != null) ? of(value) : httpGet$);
```

**Comment by supdegrave** [comment]
> as you do .pipe(first()) just after, it seems normal it does not get into
the http call

OK, maybe first() isn't the right operator to use. However, other operators
(single, skipWhile) also die as a result of the lack of complete on the
observable.

Your `switchMap` suggestion works and provides a nice cleanup, but I feel
that the concat option should be available regardless, which currently
appears to not be the case. (Also, I'm learning that with Observables there
can be numerous ways to implement things.)

**Comment by supdegrave** [comment]
Actually, it seems like the chained `switchMap` calls suffer from the same problem with missing complete. 

This code results in the complete handler in my subscribe method being called, allowing my application logic to proceed: 

```
    get$ = of(data).pipe(
        switchMap(value => value && of(value) || httpGet$)
    )
```

This code does not call the complete handler: 

```
    get$ = of(data).pipe(
        switchMap(value => value && of(value) || readFromStorage$),
        switchMap(value => value && of(value) || httpGet$)
    )
```

**Comment by Loetn** [comment]
I had the same issue with `forkJoin`. I resolved it by using `take()`, which completes the observables.

`Observable.forkJoin(observable1$.take(1), observable2$.take(1), observable3$.take(1)).subscribe(...)`.

**Comment by cyrilletuzi** [comment]
That's not normal, it's already done inside the library, so the observables complete indeed. I will investigate this next week.

**Comment by cyrilletuzi** [comment]
Fixed in 3.1.2.

Thanks for reporting.
